Ensure Operator Menu Displays All First-Level Categories

colinthebomb1 opened the pull request:
fix #3354 

This PR makes it so that when the operator menu is initialized, its minimum height will fit all first-level operator categories. The operator menu still only allows manual resizing and does not dynamically fit content because of issues with the operator menu and the result panel overlapping on smaller screens, such as laptops or phones, when dynamically fitting.

### Changes
In `left-panel.component.html`:

- Increased minimum height of resizing `left-container`
- Added template reference for `#content`

In `left-panel-component.ts`:

- Created `minWidth` and `minHeight` variables for improved readability and storing calculated minHeight
- Used `ngAfterViewInit()` to set `minHeight`  to the total height of all first-level operator categories

In `operator-menu.component.html`:

- Created `depth` variable to track first-level operator categories



### Demonstration


https://github.com/user-attachments/assets/859f5ec0-03af-4158-97ad-9cd7b9691918

GspikeHalo reviewed (commented):
Thank you so much for your PR! The dynamic effect is quite impressive. However, I believe the sidebar should maintain a fixed size rather than adjusting based on its content. 
The reason I made the title input dynamic is because there is a tag next to the input that shows the auto-save time. If the input had a fixed width, there would often be a large gap between the title and the save time, which wouldn't look good. But in this case, there isn’t a similar issue. 
To ensure consistency across all sidebars on the screen, please make sure that the size of the sidebar you modified remains fixed — just enlarge its initial size if needed. 
Thank you again!

GspikeHalo commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.ts:
```diff
@@ -21,7 +21,7 @@ export class LeftPanelComponent implements OnDestroy, OnInit {
   currentComponent: Type<any> | null = null;
   title = "Operators";
   width = 300;
```
Could we change this to 230 as well? I noticed that the width limit is set to 230 in other places too

colinthebomb1 replied:
Default width set to 230 to ensure consistency

GspikeHalo reviewed (approved):
LGTM! I left a minor suggestion. Please update the title to indicate that this PR increases the default menu size. Thanks! Please wait for Xiaozhen's review before merging this PR

Xiao-zhen-Liu commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.html:
```diff
@@ -38,7 +38,7 @@
   nz-resizable
   [nzMinWidth]="230"
-  [nzMinHeight]="300"
+  [nzMinHeight]="395"
```
I know this can fix our issue currently, but we might have more operator groups in the future and we do not want to have to change this hardcoded value each time we add a group. Based on your discussion with @GspikeHalo, For future references, it would be good to add a documentation on why we do not want to dynamically adjust the height.

colinthebomb1 replied:
After discussion, we have decided to change our solution from hardcoding the height to instead calculating a minimum height that is the the sum of all level one operator categories + padding.

Xiao-zhen-Liu commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.ts:
```diff
@@ -20,8 +20,8 @@ export class LeftPanelComponent implements OnDestroy, OnInit {
-  width = 300;
-  height = Math.max(300, window.innerHeight * 0.6);
+  width = 230;
+  height = Math.max(395, window.innerHeight * 0.6);
```
A minor suggestion: if this value is reused multiple times, can it be defined as a variable? It would make the meaning of this "395" more explicit (I assume it is a default value for the minimum height?)

colinthebomb1 replied:
I have added variable for minimum panel height and minimum panel width

Xiao-zhen-Liu reviewed (approved):
Left some minor comments. Please confirm with @GspikeHalo why dynamic resizing cannot be used here and add them to both the PR description and the code documentation. Thanks.

GspikeHalo commented:
@colinthebomb1 @Xiao-zhen-Liu 
I propose enabling manual resizing of the Operators Sidebar for the following reasons:

Previously, the sidebar height would automatically increase when a category was expanded. However, under the current configuration, the Operators Sidebar is positioned beneath the Result Panel in terms of z-index, meaning it can be visually obstructed in certain scenarios. This issue becomes especially prominent on devices with smaller screens, such as laptops.

When the sidebar dynamically expands to accommodate newly revealed content (e.g., the Visualization or Control Block sections), parts of the sidebar may become hidden behind the Result Panel. Operators located near the bottom of the sidebar may not be selectable, and even with a scrollbar, content that is visually behind the panel may remain inaccessible. As a result, users are often forced to move or minimize the Result Panel multiple times just to select an operator or interact with the sidebar—disrupting their workflow and degrading the overall user experience.

In contrast, if the sidebar supports manual resizing only, users can enlarge the visible area on demand to prevent such overlap. Even if scrolling is still required, users will always be able to access the full list of operators. Manual resizing thus improves interaction flexibility and enhances usability across different screen sizes.

That said, for improved extensibility in the future, another possible solution is **to set the minimum height of the sidebar to the total height of all first-level categories.** This way, as new categories are added, the sidebar can grow automatically without needing frequent manual updates to the layout configuration.

GspikeHalo reviewed (commented):
Please set the minimum height of the sidebar to the total height of all first-level categories

colinthebomb1 commented:
Thank you for the comments/suggestions and insight into why we use manual resizing instead of dynamically expanding the menu! I have added the logic to set the minimum height of the menu to the total height of all first-level categories. I have also added minWidth and minHeight variables to increase the readability of the code. 

I added documentation to the PR description explaining why we are not using dynamic resizing. Would you still like me to add comments to the code documenting this? If so, where should I leave these comments?

GspikeHalo commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.ts:
```diff
@@ -78,6 +81,24 @@ export class LeftPanelComponent implements OnDestroy, OnInit {
+          totalCategoriesHeight += element.clientHeight;
+        });
+
+        this.minHeight = totalCategoriesHeight + 75; // Add padding for search bar and other UI elements
```
Please extract the value `75` as a constant to make future maintenance easier. Additionally, slightly increase the bottom space of the operator sidebar so that when the user clicks the last item, the expanded content remains visible.

GspikeHalo commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.ts:
```diff
@@ -16,12 +16,15 @@ import { PanelService } from "../../service/panel/panel.service";
 })
-export class LeftPanelComponent implements OnDestroy, OnInit {
+export class LeftPanelComponent implements OnDestroy, OnInit, AfterViewInit {
+  @ViewChild("content") contentRef!: ElementRef<HTMLDivElement>;
```
Please change `contentRef ` to `content` for easier searching.

colinthebomb1 replied:
Changed contentRef to content

GspikeHalo reviewed (approved):
Thank you for your PR, it looks great! I left some minor suggestions; please make the changes.

Xiao-zhen-Liu commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.ts:
```diff
@@ -16,12 +16,15 @@ import { PanelService } from "../../service/panel/panel.service";
   title = "Operators";
-  width = 300;
-  height = Math.max(300, window.innerHeight * 0.6);
+  minWidth = 230;
```
Minor comment: usually for defining a constant class variable, it should look like this:

`private/public`(depending on which is more suitable) `static readonly MIN_PANEL_WIDTH = 230;`

I am assuming the min-width is for the left-panel component; feel free to change it to something more specific if that is not the case. Either way, you should at least:
- Be specific about what this min-width is for;
- Use ALL CAPS + underscores as the naming convention of constant class variables;
- Specify it as `static readonly` when possible.

colinthebomb1 replied:
minWidth has been changed to a private static readonly variable named MIN_PANEL_WIDTH.

Xiao-zhen-Liu commented on core/gui/src/app/workspace/component/left-panel/left-panel.component.ts:
```diff
@@ -78,6 +81,24 @@ export class LeftPanelComponent implements OnDestroy, OnInit {
     });
   }
 
+  ngAfterViewInit(): void {
```
Please add a documentation on this new design for calculating the minimum height.

colinthebomb1 replied:
Added documentation for calculating minimum height of all first-level operator categories.

Xiao-zhen-Liu reviewed (approved):
Left some minor comments. I like this new design, but please update it in the PR title, PR description, and code documentation. Also @GspikeHalo @colinthebomb1 please make sure you test it well before merging. Thanks!
